### PR TITLE
Fixed foreign-key (mariadb) errors.

### LINKED ISSUE #1050: Unable to save edits to Organization
### Subject of the issue

Organisation edits/save failure

### Your environment

* Bitwarden_rs version: 2.15.1
* Install method: docker
* Clients used: firefox
* Reverse proxy and version: traefik
* Other relevant information:

### Steps to reproduce

Create a organisation with a name and billing email address. Edit the organization and attempt to save any edits.

### Expected Behaviour

Edits are saved

### Actual Behaviour

Unable to save. Error message = "An unknown error has occured"

### Relevant logs

[CAUSE] DatabaseError(,
    "Cannot delete or update a parent row: a foreign key constraint fails (`bitwarden`.`collections`, CONSTRAINT `collections_ibfk_1` FOREIGN KEY (`org_uuid`) REFERENCES `organizations` (`uuid`))",

### LINKED ISSUE #1065: Unable to confirm email addresses on new install
### Subject of the issue
On a new setup, I am unable to configure any email addresses. The log displays the following:

```bitwarden_1    | [2020-07-21 22:41:52][bitwarden_rs::api::core::accounts][ERROR] Error saving email verification: Error saving user.
bitwarden_1    | [CAUSE] DatabaseError(
bitwarden_1    |     ForeignKeyViolation,
bitwarden_1    |     "Cannot delete or update a parent row: a foreign key constraint fails (`bitwarden`.`devices`, CONSTRAINT `devices_ibfk_1` FOREIGN KEY (`user_uuid`) REFERENCES `users` (`uuid`))",
bitwarden_1    | )
```

### Your environment

* Bitwarden_rs version: 1.16.0

* Install method: Docker
* Clients used: 
* Reverse proxy and version: Traefik 2.2
* Version of mysql/postgresql: MariaDB 10.5.4
* Other relevant information: 

### Steps to reproduce
1. Bring up Bitwarden
2. Register a new user
3. Click email verification link
4. See "Verify your email address" still visible in account
### Expected behaviour
Clicking the verification link should verify the email address

### Actual behaviour
Email address is not verified

### Relevant logs

```bitwarden_1    | [2020-07-21 22:50:18][request][INFO] POST /api/accounts/verify-email-token
bitwarden_1    | [2020-07-21 22:50:18][bitwarden_rs::api::core::accounts][ERROR] Error saving email verification: Error saving user.
bitwarden_1    | [CAUSE] DatabaseError(
bitwarden_1    |     ForeignKeyViolation,
bitwarden_1    |     "Cannot delete or update a parent row: a foreign key constraint fails (`bitwarden`.`devices`, CONSTRAINT `devices_ibfk_1` FOREIGN KEY (`user_uuid`) REFERENCES `users` (`uuid`))",
bitwarden_1    | )
bitwarden_1    | [2020-07-21 22:50:18][response][INFO] POST /api/accounts/verify-email-token (post_verify_email_token) => 200 OK```

### LINKED ISSUE #1081: MySQL backend: Cannot share cipher with organisation
### Subject of the issue
If bitwarden_rs is used with mysql backend, it is not possible to share existing ciphers with an organisation.

### Your environment
Server Installed 1.16.1
Server Latest 1.16.1
Web Installed 2.15.1
Web Latest 2.15.1

Bitwarden_rs version: In Openshift environment using docker image: bitwardenrs/server-mysql:1.16.1

### Install method
* Clients used: bitwarden app and chrome extension (irrelevant, server issue)
* Reverse proxy and version: -
* Version of mysql/postgresql: mariadb 10.5.4

### Steps to reproduce
1. Install using the mentioned docker container.
2. Create a user account
3. Create an organisation
4. Add a cipher in your user account
5. Use the share method to move it to the organisation. This will fail with "cannot save cipher"

### Expected behaviour
The cipher can be moved to the organisation.

### Actual behaviour
the cipher cannot be stored because of a constraint violation.

### Relevant logs
```
[2020-08-03 08:33:23.633][request][INFO] PUT /api/ciphers/6d708d14-07be-47f1-beee-2d2c4b83c5cd/share
[2020-08-03 08:33:23.646][error][ERROR] Error saving cipher.
[CAUSE] DatabaseError(
    ForeignKeyViolation,
    "Cannot delete or update a parent row: a foreign key constraint fails (`bitwarden`.`ciphers_collections`, CONSTRAINT `ciphers_collections_ibfk_1` FOREIGN KEY (`cipher_uuid`) REFERENCES `ciphers` (`uuid`))",
)
[2020-08-03 08:44:54.592][response][INFO] PUT /api/ciphers/<uuid>/share (put_cipher_share) => 400 Bad Request
```
If the contraint is divided in update and delete constraints, it turns out that the delete contraint is violated:

```
[2020-08-03 08:53:44.893][request][INFO] PUT /api/ciphers/10d228e6-4f28-4cba-a4fd-52944646fc04/share
[2020-08-03 08:53:44.907][error][ERROR] Error saving cipher.
[CAUSE] DatabaseError(
    ForeignKeyViolation,
    "Cannot delete or update a parent row: a foreign key constraint fails (`bitwarden`.`ciphers_collections`, CONSTRAINT `delete_check` FOREIGN KEY (`cipher_uuid`) REFERENCES `ciphers` (`uuid`) ON UPDATE NO ACTION)",
)
[2020-08-03 08:53:44.907][response][INFO] PUT /api/ciphers/<uuid>/share (put_cipher_share) => 400 Bad Request
````

It seems that the server wants to recreate the cipher in the database instead of updating it. By trying that, it violated the key constraints of the *ciphers_collections*.
It apperas that the *folder_collection* is also affected.

### LINKED ISSUE #1159: Error of edit Organization information
When I edit organization information like Business name/Identifier, it shows "An error has occurred.
Error saving organization". 
The bitwarden log:
is [error][ERROR] Error saving organization.
[CAUSE] DatabaseError(
    ForeignKeyViolation,
    "Cannot delete or update a parent row: a foreign key constraint fails (`bitwarden`.`ciphers`, CONSTRAINT `ciphers_ibfk_2` FOREIGN KEY (`organization_uuid`) REFERENCES `organizations` (`uuid`))",
)

### PR DESCRIPTION
**Pull request by BlackDex**
When using MariaDB v10.5+ Foreign-Key errors were popping up because of
some changes in that version. To mitigate this on MariaDB and other
MySQL forks those errors are now catched, and instead of a replace_into
an update will happen. I have tested this as thorough as possible with
MariaDB 10.5, 10.4, 10.3 and the default MySQL on Ubuntu Focal. And
tested it again using sqlite, all seems to be ok on all tables.

resolves #1081. resolves #1065, resolves #1050, resolves #1159

### CODE REVIEW

File: src/db/models/attachment.rs [hunk 1]
```diff
@@ -63,10 +63,21 @@ impl Attachment {
...
         db_run! { conn:
             sqlite, mysql {
-                diesel::replace_into(attachments::table)
+                match diesel::replace_into(attachments::table)
                     .values(AttachmentDb::to_db(self))
                     .execute(conn)
-                    .map_res("Error saving attachment")
+                {
+                    Ok(_) => Ok(()),
+                    // Record already exists and causes a Foreign Key Violation because replace_into() want's to delete the record first.
```
Comment (jjlin) [inline src/db/models/attachment.rs:71]: `want's` should be `wants`.

File: src/db/models/collection.rs [hunk 2]
```diff
@@ -344,8 +372,9 @@ impl CollectionCipher {
     pub fn save(cipher_uuid: &str, collection_uuid: &str, conn: &DbConn) -> EmptyResult {
         Self::update_users_revision(&collection_uuid, conn);
 
-        db_run! { conn: 
+        db_run! { conn:
             sqlite, mysql {
+                // Not checking for ForeignKey Constraints here
```
Comment (jjlin) [inline src/db/models/collection.rs:377]: It might be helpful to mention why not

File: src/db/models/folder.rs [hunk 3]
```diff
@@ -136,7 +147,7 @@ impl Folder {
 
 impl FolderCipher {
     pub fn save(&self, conn: &DbConn) -> EmptyResult {
-        db_run! { conn: 
+        db_run! { conn:
             sqlite, mysql {
                 diesel::replace_into(folders_ciphers::table)
```
Comment (jjlin) [inline src/db/models/folder.rs:152]: Same here

File: src/db/models/user.rs [hunk 4]
```diff
@@ -290,18 +301,19 @@ impl Invitation {
 
         db_run! {conn:
             sqlite, mysql {
+                // Not checking for ForeignKey Constraints here
```
Comment (jjlin) [inline src/db/models/user.rs:304]: And here

**Comment by BlackDex** [comment]
@jjlin Made the changes, thanks for the remarks.

File: src/db/models/folder.rs [hunk 5]
```diff
@@ -136,8 +147,10 @@ impl Folder {
 
 impl FolderCipher {
     pub fn save(&self, conn: &DbConn) -> EmptyResult {
-        db_run! { conn: 
+        db_run! { conn:
             sqlite, mysql {
+                // Not checking for ForeignKey Constraints here.
+                // If they do occure then this record was already located in the table which is fine.
```
Comment (jjlin) [inline src/db/models/folder.rs:153]: I think the main thing here, as with the other `replace_into()` instances that can't have foreign key constraint errors, is that `folders_ciphers` doesn't have any fields that are referenced by other tables (as foreign keys).

Comment (BlackDex) [reply src/db/models/folder.rs:153]: They do. The folder uuid and cipher uuid are both referenced. But in the case they are together already it doesn't matter in this case.

Comment (jjlin) [reply src/db/models/folder.rs:153]: `folders.uuid` and `ciphers.uuid` are referenced as foreign keys by various tables (including `folders_ciphers`), but `folders_ciphers.cipher_uuid` and `folders_ciphers.folder_uuid` are not.

Comment (BlackDex) [reply src/db/models/folder.rs:153]: Ill check it again tomorrow, i thought I saw that but could be the other way around indeed.

Comment (BlackDex) [reply src/db/models/folder.rs:153]: I could also remove the comment of course ;)

Comment (jjlin) [reply src/db/models/folder.rs:153]: Well, there is some subtlety to the point, so I think it's worthwhile to add more explanation, so long as it's accurate

Comment (BlackDex) [reply src/db/models/folder.rs:153]: The last remark was sarcasm ;).

But indeed, the ERD showed it in a weird way, looking at the DDL it self indeed doesn't mention that. I will change the comments to better represent this.

**Comment by BlackDex** [comment]
@jjlin I explained it a bit better i hope.
Else, if you have any suggests, please share them :).

**Comment by dani-garcia** [comment]
Thanks!

**Pull request merged**